web/models/pair_retro.ex

```diff
@@ -0,0 +1,70 @@
+defmodule Pairmotron.PairRetro do
+  use Pairmotron.Web, :model
+
+  schema "pair_retros" do
+    field :pair_date, Ecto.Date
+    field :subject, :string
+    field :reflection, :string
+    belongs_to :user, Pairmotron.User
+    belongs_to :pair, Pairmotron.Pair
+    belongs_to :project, Pairmotron.Project
+
+    timestamps()
+  end
+
+  @required_fields ~w(pair_date user_id pair_id)
+  @optional_fields ~w(subject reflection project_id)
+
+  @doc """
+  Builds a changeset based on the `struct` and `params`.
+
+  pair_start_date should be a date which is the first day
+  of the week and year on the associated pair. It is used
+  to validate that the pair_date of this pair_retro is after
+  that date, since the actual pairing could not have occurred
+  before the actual pair was assigned.
+  """
+  def changeset(struct, params \\ %{}, pair_start_date) do
+    struct
+    |> cast(params, @required_fields, @optional_fields)
+    |> foreign_key_constraint(:user_id)
+    |> foreign_key_constraint(:pair_id)
+    |> foreign_key_constraint(:project_id)
+    |> validate_field_is_not_before_date(:pair_date, pair_start_date)
+    |> validate_field_is_not_in_future(:pair_date)
+  end
```

Comment:
So the decision was made based on conversations with @ericworkman and with denizens of the Ecto room of the elixir-lang slack to pass in an additional parameter to the changeset (`pair_start_date`) rather than the alternative of making a database call from within the `validate_field_is_not_before_date` changeset validator. I'm on the fence about it.

On the side of passing the `pair_start_date` into the changeset, it isolated all of the database interaction to the controller. This injects the pair_start_date dependency into the changeset function, thereby confining the database lookup side effect to the controller.

There is an argument for not passing the pair_start_date parameter into the changeset at all, and instead invoking `Repo.get(Pairmotron.Pair, pair_id)` from within the `validate_field_is_not_before_date` function to retrieve the `pair_start_date`. The database call would be called through `Ecto.Changeset.validate_change`, which only invokes the logic if the given field `:pair_date` actually changes. This eliminates a database select when the field isn't actually changed. This logic (eliminating unnecessary database calls) could be duplicated in the controller, but it would be duplicated functionality, and would add complexity to the controller. The downside of this approach is making the `validate_field_is_not_before_date` less abstract, and also impure.